### PR TITLE
Support tf 0.13 for eks

### PR DESCRIPTION
**Pull request by scottaubrey**
Changes necessary to get terraform plan to be successful and clean on tf 0.13

### CODE REVIEW

**Comment by scottaubrey** [comment]
This (combined with changes [here](https://github.com/elifesciences/builder/pull/1100/files)) results in a plan of:
```
INFO - cfn - Create: {'Outputs': {}, 'Parameters': {}, 'Resources': {}}
INFO - cfn - Update: {'Outputs': {}, 'Resources': {}}
INFO - cfn - Delete: {'Outputs': {}, 'Parameters': {}, 'Resources': {}}
INFO - cfn - Terraform delta: 
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place
 <= read (data resources)
Terraform will perform the following actions:
  # data.tls_certificate.oidc_cert will be read during apply
  # (config refers to values not yet known)
 <= data "tls_certificate" "oidc_cert"  {
        certificates = [
            {
                is_ca                = true
                issuer               = "OU=Starfield Class 2 Certification Authority,O=Starfield Technologies\\, Inc.,C=US"
                not_after            = "2034-06-28T17:39:16Z"
                not_before           = "2009-09-02T00:00:00Z"
                public_key_algorithm = "RSA"
                serial_number        = "12037640545166866303"
                sha1_fingerprint     = "9e99a48a9960b14926bb7f3b02e22da2b0ab7280"
                signature_algorithm  = "SHA256-RSA"
                subject              = "CN=Starfield Services Root Certificate Authority - G2,O=Starfield Technologies\\, Inc.,L=Scottsdale,ST=Arizona,C=US"
                version              = 3
            },
            {
                is_ca                = true
                issuer               = "CN=Starfield Services Root Certificate Authority - G2,O=Starfield Technologies\\, Inc.,L=Scottsdale,ST=Arizona,C=US"
                not_after            = "2037-12-31T01:00:00Z"
                not_before           = "2015-05-25T12:00:00Z"
                public_key_algorithm = "RSA"
                serial_number        = "144918191876577076464031512351042010504348870"
                sha1_fingerprint     = "06b25927c42a721631c1efd9431e648fa62e1e39"
                signature_algorithm  = "SHA256-RSA"
                subject              = "CN=Amazon Root CA 1,O=Amazon,C=US"
                version              = 3
            },
            {
                is_ca                = true
                issuer               = "CN=Amazon Root CA 1,O=Amazon,C=US"
                not_after            = "2030-08-23T22:21:28Z"
                not_before           = "2022-08-23T22:21:28Z"
                public_key_algorithm = "RSA"
                serial_number        = "166129328851546858514271303855646110030630415"
                sha1_fingerprint     = "2ad974a775f73cbdbbd8f5ac3a49255fa8fb1f8c"
                signature_algorithm  = "SHA256-RSA"
                subject              = "CN=Amazon RSA 2048 M01,O=Amazon,C=US"
                version              = 3
            },
            {
                is_ca                = false
                issuer               = "CN=Amazon RSA 2048 M01,O=Amazon,C=US"
                not_after            = "2023-06-26T23:59:59Z"
                not_before           = "2023-02-23T00:00:00Z"
                public_key_algorithm = "RSA"
                serial_number        = "12963317977521044204933868342894944555"
                sha1_fingerprint     = "f8e9efef11f4308c0912577ed37afe2707b2721c"
                signature_algorithm  = "SHA256-RSA"
                subject              = "CN=oidc.eks.us-east-1.amazonaws.com"
                version              = 3
            },
        ]
      ~ id           = "2023-04-05 14:20:08.834621 +0000 UTC" -> "2023-04-05 14:20:15.004311 +0000 UTC"
        url          = "https://oidc.eks.us-east-1.amazonaws.com/id/C44C02779528D1AE873C5052FC630D5C"
        verify_chain = true
    }
  # aws_autoscaling_group.worker will be updated in-place
  ~ resource "aws_autoscaling_group" "worker" {
        arn                       = "arn:aws:autoscaling:us-east-1:512686554592:autoScalingGroup:0033013d-6760-40d5-b909-1a6a410ecf67:autoScalingGroupName/kubernetes-aws--flux-test--worker"
        availability_zones        = [
            "us-east-1c",
            "us-east-1d",
        ]
        default_cooldown          = 300
        desired_capacity          = 2
        enabled_metrics           = []
        force_delete              = false
        health_check_grace_period = 300
        health_check_type         = "EC2"
        id                        = "kubernetes-aws--flux-test--worker"
        launch_configuration      = "kubernetes-aws--flux-test--worker20230404075008513000000001"
        load_balancers            = []
        max_size                  = 6
        metrics_granularity       = "1Minute"
        min_size                  = 1
        name                      = "kubernetes-aws--flux-test--worker"
        protect_from_scale_in     = false
        service_linked_role_arn   = "arn:aws:iam::512686554592:role/aws-service-role/autoscaling.amazonaws.com/AWSServiceRoleForAutoScaling"
        suspended_processes       = []
      ~ tags                      = [
          - {
              - "key"                 = "Project"
              - "propagate_at_launch" = "1"
              - "value"               = "kubernetes-aws"
            },
          - {
              - "key"                 = "Environment"
              - "propagate_at_launch" = "1"
              - "value"               = "flux-test"
            },
          - {
              - "key"                 = "Name"
              - "propagate_at_launch" = "1"
              - "value"               = "kubernetes-aws--flux-test"
            },
          - {
              - "key"                 = "Cluster"
              - "propagate_at_launch" = "1"
              - "value"               = "kubernetes-aws--flux-test"
            },
          - {
              - "key"                 = "kubernetes.io/cluster/kubernetes-aws--flux-test"
              - "propagate_at_launch" = "1"
              - "value"               = "owned"
            },
          + {
              + "key"                 = "Project"
              + "propagate_at_launch" = "true"
              + "value"               = "kubernetes-aws"
            },
          + {
              + "key"                 = "Environment"
              + "propagate_at_launch" = "true"
              + "value"               = "flux-test"
            },
          + {
              + "key"                 = "Name"
              + "propagate_at_launch" = "true"
              + "value"               = "kubernetes-aws--flux-test"
            },
          + {
              + "key"                 = "Cluster"
              + "propagate_at_launch" = "true"
              + "value"               = "kubernetes-aws--flux-test"
            },
          + {
              + "key"                 = "kubernetes.io/cluster/kubernetes-aws--flux-test"
              + "propagate_at_launch" = "true"
              + "value"               = "owned"
            },
        ]
        target_group_arns         = []
        termination_policies      = []
        vpc_zone_identifier       = [
            "subnet-0110e7685debdb5eb",
            "subnet-07e912335040a0866",
        ]
        wait_for_capacity_timeout = "10m"
        timeouts {}
    }
Plan: 0 to add, 1 to change, 0 to destroy.
```

I'm not sure on the ID changing for the cert data, but that is just a data provider, isn't used and has no reference anywhere else - I'm not worried.

The autoscaling group tags look innocuous enough, changing `propagate_at_launch` from a `"1"` to a `"true"`. The necessitates removing all the tags and re-adding them. I don't imagine this is a problem and it only applies to new instances anyway. So we could apply, then test it straight away. If it's the `propagate_at_launch` flag isn't set correctly in someway, I can change the value [here](https://github.com/elifesciences/builder/blob/72ff4e278c9fb9969b4a9041665a2b5c181c19bc/src/buildercore/terraform.py#L987) to "1" and reapply.

In otherwords, I think with the caveat above, that we're good to go with merging this, and applying (and updating state) of the `kubernetes-aws` project.

**Comment by lsh-0** [comment]
> I'm not sure on the ID changing for the cert data, but that is just a data provider, isn't used and has no reference anywhere else - I'm not worried.

I'm not seeing any ID changing for the cert data. From the output I see:

```
Resource actions are indicated with the following symbols:
  ~ update in-place
 <= read (data resources)
```

and then

```
Terraform will perform the following actions:
 # data.tls_certificate.oidc_cert will be read during apply
 # (config refers to values not yet known)
 <= data "tls_certificate" "oidc_cert"  {
```

So it looks like it's being left alone.

> In otherwords, I think with the caveat above, that we're good to go with merging this, and applying (and updating state) of the kubernetes-aws project.

I agree, go for it.

I _am_ surprised you weren't bitten by `providers` requiring a `required_providers` block under `terraform` though: https://github.com/elifesciences/builder/pull/1097

It could be because `aws` is the default `hashicorp/aws` provider? No idea.

**Review by lsh-0 — APPROVED** [review]

**Comment by lsh-0** [comment]
oh, I see it now:

>       ~ id           = "2023-04-05 14:20:08.834621 +0000 UTC" -> "2023-04-05 14:20:15.004311 +0000 UTC"

Missed it before. Yeah. Weird. It's gained 7 seconds? Or 7 seconds since you last ran the command? It could be a nuisance in the future if it becomes a perpetual diff but that sounds like a bug.

File: src/buildercore/terraform.py
Comment (scottaubrey) [inline src/buildercore/terraform.py:1]: @lsh-0 FYI - it works for me, but not sure if it's right or not. The issue I was having yesterday morning was:
```
exception while executing task 'update_infrastructure': 'provider-kubernetes'

Traceback (most recent call last):
  File "src/taskrunner.py", line 283, in exec_task
    return_map['result'] = task_map['fn'](*task_args, **task_kwargs)
  File "/Volumes/Sources/elife/builder/src/decorators.py", line 19, in wrap
    result = fn(*args, **kw)
  File "/Volumes/Sources/elife/builder/src/cfn.py", line 80, in update_infrastructure
    context, delta, current_context = cfngen.regenerate_stack(stackname, **more_context)
  File "/Volumes/Sources/elife/builder/src/buildercore/cfngen.py", line 1079, in regenerate_stack
    delta = template_delta(context)
  File "/Volumes/Sources/elife/builder/src/buildercore/cfngen.py", line 1041, in template_delta
    terraform.generate_delta(context)
  File "/Volumes/Sources/elife/builder/src/buildercore/terraform.py", line 1390, in generate_delta
    return plan(new_context)
  File "/Volumes/Sources/elife/builder/src/buildercore/terraform.py", line 1393, in plan
    terraform = init(context['stackname'], context)
  File "/Volumes/Sources/elife/builder/src/buildercore/terraform.py", line 1549, in init
    required_providers = dict(map(provider_to_required_provider, providers['provider']))
  File "/Volumes/Sources/elife/builder/src/buildercore/terraform.py", line 1543, in provider_to_required_provider
    source = context['terraform'][provider_context_key].get('source', default_source)
KeyError: 'provider-kubernetes'
```

Comment (lsh-0) [reply src/buildercore/terraform.py:1]: this is helpful, I'll take a quick look before I sign off

Comment (scottaubrey) [reply src/buildercore/terraform.py:1]: i.e. [this code block](https://github.com/elifesciences/builder/blob/master/src/buildercore/terraform.py#L1482-L1496) is where 'kubernetes" is defined

**Pull request merged**